### DISCUSSION TITLE
Annotation for native Java functions

### DISCUSSION THREAD

**Discussion by Jongy** [async-profiler/async-profiler#665] [Ideas]
`native` Java functions, i.e a function like `private native Class<?> defineClass1(String name, byte[] b, int off, int len, ProtectionDomain pd, String source);`, will appear in the profile result with `_[j]` suffix followed by the native implementation `Java_java_lang_ClassLoader_defineClass1`.
Can it be marked differently? e.g `_[n]`. I think it will make the profile easier to read, when there are mixed native functions called by Java & native functions calling back to Java via JNI. I don't know if AGCT provides this information though.

**Comment by apangin** [comment]
Do you mean marking all Java `native` methods?
It's easy to implement, but I'm not sure about the benefits.
Native function can be also called from the interpreter or from a compiled native wrapper. This information is currenty encoded with either `_[0]` or `_[j]` suffix, like for regular Java methods.
What frame type do you expect these methods to have in JFR format?

**Reply by Jongy** [reply]
I'm referring to the Java frame created by async-profiler. Let's take this function for example: `static native int epollWait(int epfd, long pollAddress, int numfds) IOException` from `sun/nio/ch/EPollArrayWrapper`. The profiler creates a frame for it `sun/nio/ch/EPollArrayWrapper.epollWait(IJI)I`. This frame, to my understanding, will always be marked with a `_[j]` suffix in the `collapsed` format. This function is called by JNI and the native function called is `Java_sun_nio_ch_EPollArrayWrapper_epollWait`.

So what I'm asking is, can the `sun/nio/ch/EPollArrayWrapper.epollWait(IJI)I` frame be marked with something else, not `_[j]`, so it can be determined as a "native wrapper" and not really Java code. Maybe in JFR this is already the case? (I'm checking against `collapsed`)

**Reply by apangin** [reply]
> This frame, to my understanding, will always be marked with a _[j] suffix in the collapsed format.

As I mentioned above, the suffix is either `_[0]` or `_[j]`, depending on whether the JVM executes an interpreter native_entry or JIT-compiled native_wrapper. Despite lacking Java bytecode, native methods do run some generated machine code before entering an associated native function.

Flight Recorder marks frames of Java native methods with `Native` type. This does not make much sense to me, since native methods already have `ACC_NATIVE` flag set in JFR constant pool.
async-profiler uses `Native` type to mark non-Java code (which Flight Recorder does not show at all).

What I'm saying, even though adding `_[n]` suffix looks trivial to implement (and it actually is), this raises many consequent questions like:
 - should we add more suffixes to distinguish interpreted/compiled wrappers;
 - whether we need to add more JFR frame types (multiplied by the number of JIT statuses) - this in turn may affect JMC;
 - do we need to add more Flame Graph colors (there are a lot of them already)?

Considering the consequences, this no longer seems a trivial change, so I'd like to understand better what benefits this will bring instead.

**Reply by Jongy** [reply]
Ah, I didn't understand you earlier (didn't know that the "native wrapper" for a JNI method could itself be interpreted or JITted).
I don't have any concrete idea now on the benefits this can get, but my approach in general is to have async-profiler provide as much meta-data on frames as possible, so different traits such as compilation level / native / are useful to be included.

We can close the question, your answer explain it to me, now what's left is a feature request that is done separately :) Thanks Andrei!
